core/src/com/unciv/ui/civilopedia/CivilopediaText.kt

```diff
@@ -111,3 +177,106 @@ object MarkupRenderer {
+ * 4. Optionally override [getCivilopediaTextLines] to supply automatic stuff like tech prerequisites, uniques, etc.
+ * 4. Optionally override [assembleCivilopediaText] to handle assembly of the final set of lines yourself.
+ */
+interface ICivilopediaText {
+    /** List of strings supporting simple [formatting rules][FormattedLine] that [CivilopediaScreen] can render.
+     * May later be merged with automatic lines generated by the deriving class
+     *  through overridden [getCivilopediaTextHeader] and/or [getCivilopediaTextLines] methods.
+     *
+     */
+    var civilopediaText: List<FormattedLine>
+
+    /** Generate header line from object metadata.
+     * Default implementation will pull [INamed.name] and render it in 150% normal font size.
+     * @return A [FormattedLine] that will be inserted on top
+     */
+    fun getCivilopediaTextHeader(): FormattedLine? =
+        if (this is INamed) FormattedLine(name, header = 2)
+        else null
+
+    /** Generate automatic lines from object metadata.
+     *
+     * Default implementation is empty - no need to call super in overrides.
+     *
+     * @param ruleset The current ruleset for the Civilopedia viewer
+     * @return A list of [FormattedLine]s that will be inserted before
+     *         the first line of [civilopediaText] having a [link][FormattedLine.link]
+     */
+    fun getCivilopediaTextLines(ruleset: Ruleset): List<FormattedLine> = listOf()
+
+    /** Override this and return true to tell the Civilopedia that the legacy description is no longer needed */
+    fun replacesCivilopediaDescription() = false
+    /** Override this and return true to tell the Civilopedia that this is not empty even if nothing came from json */
+    fun hasCivilopediaTextLines() = false
+    /** Indicates that neither json nor getCivilopediaTextLines have content */
+    fun isCivilopediaTextEmpty() = civilopediaText.isEmpty() && !hasCivilopediaTextLines()
+
+    /** Build a Gdx [Table] showing our [formatted][FormattedLine] [content][civilopediaText]. */
+    fun renderCivilopediaText (labelWidth: Float, linkAction: ((id: String)->Unit)? = null): Table {
+        return MarkupRenderer.render(civilopediaText, labelWidth, linkAction = linkAction)
+    }
+
+    /** Assemble json-supplied lines with automatically generated ones.
+     *
+     * The default implementation will insert [getCivilopediaTextLines] before the first [linked][FormattedLine.link] [civilopediaText] line and [getCivilopediaTextHeader] on top.
+     *
+     * @param ruleset The current ruleset for the Civilopedia viewer
+     * @return A new CivilopediaText instance containing original [civilopediaText] lines merged with those from [getCivilopediaTextHeader] and [getCivilopediaTextLines] calls.
+     */
+    fun assembleCivilopediaText(ruleset: Ruleset): CivilopediaText {
+        val outerLines = civilopediaText.iterator()
+        val newLines = sequence {
```

Comment:
Oh, nice! Clever use of yield!

Reply:
Still one of the parts I wasn't too happy with - but I wanted a specific ability, namely mixing json lines with generated lines in a grouped kind of order, so it came out much longer than I'd have preferred.